packages/jupyter-ai/jupyter_ai/history.py

```diff
@@ -0,0 +1,40 @@
+from typing import List, Sequence
+
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import BaseMessage
+from langchain_core.pydantic_v1 import BaseModel, Field
+
+
+class BoundedChatHistory(BaseChatMessageHistory, BaseModel):
+    """
+    An in-memory implementation of `BaseChatMessageHistory` that stores up to
+    `k` exchanges between a user and an LLM.
+
+    For example, when `k=2`, `BoundedChatHistory` will store up to 2 human
+    messages and 2 AI messages.
+    """
+
+    messages: List[BaseMessage] = Field(default_factory=list)
+    size: int = 0
+    k: int
+
+    async def aget_messages(self) -> List[BaseMessage]:
+        return self.messages
+
+    def add_message(self, message: BaseMessage) -> None:
+        """Add a self-created message to the store"""
+        self.messages.append(message)
+        self.size += 1
+
+        if self.size > self.k * 2:
+            self.messages.pop(0)
+
+    async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
```

Comment:
Is it common Python practice to name a method `foo` if it's synchronous and `afoo` if it's async? `aadd_messages` read like a typo at first, but if this naming convention is typical, then let's leave it as it is.

Reply:
I don't think there is a generally-accepted naming convention to distinguish sync methods from async methods. However, we have to define the method as `aadd_messages()` because that is expected to be available when used as a history object by LangChain. This can be verified from the `BaseChatMessageHistory` docs: https://api.python.langchain.com/en/latest/chat_history/langchain_core.chat_history.BaseChatMessageHistory.html#